docs/language/ql-training-rst/java/program-representation-java.rst

```diff
@@ -0,0 +1,167 @@
+   
+     try {
+         ...
+     } catch (AnException e) {
+     }
+
+.. container:: ast-graph
+     
+      .. graphviz::
+         
+            digraph {
+            graph [ dpi = 1000 ]
+            node [shape=polygon,sides=4,color=blue4,style="filled,rounded",   fontname=consolas,fontcolor=white]
+            a [label=<TryStmt>]
+            b [label=<CatchClause>]
+            c [label=<...>,color=white,fontcolor=black]
+            d [label=<LocalVariable<BR />DeclExpr>]
+            e [label=<...>,color=white,fontcolor=black]
+            f [label=<...>,color=white,fontcolor=black]
+            g [label=<...>,color=white,fontcolor=black]
+   
+            a -> {b, c}
+            b -> {d, e}
+            d -> {f, g}
+         }
+
+
+
+.. note::
+
+  When writing queries in QL it is important to have in mind the underlying representation of the program which is stored in the database. Typically queries make use of the “AST” representation of the program - a tree structure where program elements are nested within other program elements.
+
+  The “Introducing the Java libraries” help topic contains a more complete overview of important AST classes and the rest of the Java QL libraries: https://help.semmle.com/QL/learn-ql/ql/java/introduce-libraries-java.html 
+
+Database representations of ASTs
+================================
+
+AST nodes and other program elements are encoded in the database as *entity values*. Entities are implemented as integers, but in QL they are opaque - all one can do with them is to check their equality.
+
+Each entity belongs to an entity type. Entity types have names starting with “@” and are defined in the database schema (not in QL).
+
+Properties of AST nodes and their relationships to each other are encoded by database relations, which are predicates defined in the database (not in QL).
+
+Entity types are rarely used directly, the usual pattern is to define a QL class that extends the type and exposes properties of its entities through member predicates.
+
+.. note::
+
+  ASTs are a typical example of the kind of data representation one finds in object-oriented programming, with data-carrying nodes that reference each other. At first glance, QL, which can only work with atomic values, does not seem to be well suited for working with this kind of data. However, ultimately all that we require of the nodes in an AST is that they have an identity. The relationships among nodes, usually implemented by reference-valued object fields in other languages, can just as well (and arguably more naturally) be represented as relations over nodes. Attaching data (such as strings or numbers) to nodes can also be represented with relations over nodes and primitive values. All we need is a way for relations to reference nodes. This is achieved in QL (as in other database languages) by means of *entity values* (or entities, for short), which are opaque atomic values, implemented as integers under the hood.
```

Comment:
This bunch of notes seems pretty generic until the final sentence. Also, it seems likely that it is closely related to text in an existing topic. Possibly this is a candidate for the "snippet" library.